eradiate/radprops/vert_dist.py

```diff
@@ -0,0 +1,409 @@
+"""
+Particle vertical distributions.
+"""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Callable
+
+import attr
+import numpy as np
+import pint
+import pinttr
+import xarray as xr
+from pinttr.util import units_compatible
+from scipy.stats import expon, norm
+
+from .._attrs import documented, parse_docs
+from .._factory import BaseFactory
+from ..units import to_quantity
+from ..units import unit_context_config as ucc
+from ..units import unit_registry as ureg
+
+
+@parse_docs
+@attr.s
+class VerticalDistribution(ABC):
+    r"""
+    An abstract base class for particles vertical distributions.
+
+    Vertical distributions help define particle layers.
+
+    The particle layer is split into a number of divisions (sub-layers),
+    wherein the particles fraction is evaluated.
+
+    The vertical distribution is normalised so that:
+
+    .. math::
+        \sum_i f_i = 1
+
+    where :math:`f_i` is the particles fraction in the layer division :math:`i`.
+    """
+    bottom = documented(
+        pinttr.ib(
+            units=ucc.deferred("length"),
+            converter=pinttr.converters.to_units(ucc.deferred("length")),
+            validator=pinttr.validators.has_compatible_units,
+        ),
+        doc="Layer bottom altitude.\n\nUnit-enabled field (default: ucc[length]).",
+        type="float",
+    )
+    top = documented(
+        pinttr.ib(
+            units=ucc.deferred("length"),
+            converter=pinttr.converters.to_units(ucc.deferred("length")),
+            validator=pinttr.validators.has_compatible_units,
+        ),
+        doc="Layer top altitude.\n\nUnit-enabled field (default: ucc[length]).",
+        type="float",
+    )
+
+    def __attrs_post_init__(self):
+        if self.bottom >= self.top:
+            raise ValueError("bottom altitude must be lower than top altitude")
+
+    @classmethod
+    def from_dict(cls, d: dict) -> VerticalDistribution:
+        """
+        Initialise a :class:`VerticalDistribution` from a dictionary.
+        """
+        return cls(**d)
+
+    @property
+    @abstractmethod
+    def fractions(self) -> Callable:
+        """
+        Returns a callable that evaluates the particles fractions in the
+        layer, given an array of altitude values.
+        """
+        pass
+
+    @staticmethod
+    def _normalise(x: np.ndarray) -> np.ndarray:
+        """
+        Scale the values so that their sum is 1.
+
+        Parameter ``x`` (array):
+            Values to normalise.
+
+        Returns → array:
+            Normalised array.
+        """
+        _norm = np.sum(x)
+        if _norm > 0.0:
+            return x / _norm
+        else:
+            raise ValueError(f"Cannot normalise fractions because the norm is " f"0.")
```

Comment:
If I don't turn these 4 lines of code into a function, I have to repeat them in each `fractions` property of each `VerticalDistribution` class, don't I?

Reply:
My point is: you don't need 4 lines, because `ZeroDivisionError` is explicit. So just writing `foo = x / np.sum(x)` achieves the same. Can actually `np.sum(x)` be negative?

Reply:
Actually, you have an [even more concise option](https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html): `foo = x / np.linalg.norm(x, ord=0)`.